RecoLocalTracker/SiPixelClusterizer/plugins/gpuCalibPixel.h

```diff
@@ -64,6 +69,50 @@ namespace gpuCalibPixel {
       }
     }
   }
+  __global__ void calibDigisUpgrade(uint16_t* xx,
+                                    uint16_t* yy,
+                                    uint16_t* adc,
+                                    uint32_t* pdigi,
+                                    uint16_t* id,
+                                    int numElements,
+                                    uint32_t* __restrict__ moduleStart,        // just to zero first
+                                    uint32_t* __restrict__ nClustersInModule,  // just to zero them
+                                    uint32_t* __restrict__ clusModuleStart) {
+    int first = blockDim.x * blockIdx.x + threadIdx.x;
+
+    if (0 == first)
+      clusModuleStart[0] = moduleStart[0] = 0;
+    for (int i = first; i < gpuClustering::MaxNumModulesUpgrade; i += gridDim.x * blockDim.x) {
+      nClustersInModule[i] = 0;
+    }
+
+    for (int i = first; i < numElements; i += gridDim.x * blockDim.x) {
+      if (InvId == id[i])
+        continue;
+      int mode = (Phase2ReadoutMode < -1 ? -1 : Phase2ReadoutMode);
+
+      if (mode < 0)
+        adc[i] = std::max(100, int(adc[i] * ElectronPerADCGain));
+      else {
+        if (adc[i] < Phase2KinkADC)
+          adc[i] = int((adc[i] - 0.5) * ElectronPerADCGain);
```

Comment:
@AdrianoDee as discussed today the [phase-2 simulation meeting](https://indico.cern.ch/event/949831/contributions/4148477/attachments/2165736/3655168/IT_local_Reco_GPU_adriano_dic2020.pdf), you are indeed missing here the update the in pixel clusterizer due to the bug-fix applied when turning on the dual-slope ToT here:

https://github.com/cms-sw/cmssw/pull/31253/commits/0cfe9f7a1066a0f89de56081faac65a24bd9209d#

This might explain why the GPU results (in which you calibrate the digis) are worse than the CPU ones (in which you take predigested clusters from the legacy reco).

@emiglior @tsusa

Reply:
@mmusich thanks! Updating it accordingly to cms-sw@0cfe9f7